x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/utils/ToXContentCompressor.java

```diff
@@ -0,0 +1,78 @@
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.core.ml.inference.utils;
+
+import org.elasticsearch.common.CheckedFunction;
+import org.elasticsearch.common.bytes.BytesArray;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.io.Streams;
+import org.elasticsearch.common.io.stream.BytesStreamOutput;
+import org.elasticsearch.common.xcontent.LoggingDeprecationHandler;
+import org.elasticsearch.common.xcontent.NamedXContentRegistry;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+import org.elasticsearch.common.xcontent.XContentHelper;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.XContentType;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import java.util.Map;
+import java.util.zip.GZIPInputStream;
+import java.util.zip.GZIPOutputStream;
+
+/**
+ * Collection of helper methods. Similar to CompressedXContent, but this utilizes GZIP for parity with the native compression
+ */
+public final class ToXContentCompressor {
+    private static final int BUFFER_SIZE = 4096;
+
+    private ToXContentCompressor() {}
+
+    public static <T extends ToXContentObject> String deflate(T objectToCompress) throws IOException {
+        BytesReference reference = XContentHelper.toXContent(objectToCompress, XContentType.JSON, false);
+        return deflate(reference);
+    }
+
+    public static <T> T inflate(String compressedString,
+                                CheckedFunction<XContentParser, T, IOException> parserFunction,
+                                NamedXContentRegistry xContentRegistry) throws IOException {
+        try(XContentParser parser = XContentHelper.createParser(xContentRegistry,
+            LoggingDeprecationHandler.INSTANCE,
+            inflate(compressedString),
+            XContentType.JSON)) {
+            return parserFunction.apply(parser);
+        }
+    }
+
+    public static Map<String, Object> inflateToMap(String compressedString) throws IOException {
+        // Don't need the xcontent registry as we are now deflating named objects.
+        try(XContentParser parser = XContentHelper.createParser(NamedXContentRegistry.EMPTY,
+            LoggingDeprecationHandler.INSTANCE,
+            inflate(compressedString),
+            XContentType.JSON)) {
+            return parser.mapOrdered();
+        }
+    }
+
+    static BytesReference inflate(String compressedString) throws IOException {
+        byte[] compressedBytes = Base64.getDecoder().decode(compressedString.getBytes(StandardCharsets.UTF_8));
+        InputStream inflateStream = new GZIPInputStream(new BytesArray(compressedBytes).streamInput(), BUFFER_SIZE);
```

Comment:
I think a `BoundedInputStream` (apache commons) should be used to protect for zip bombs. The generic ToMap parser can be abused to cause a OOM and crash the node.